_posts/2017-10-06-rust-blog-series-2.md

```diff
@@ -0,0 +1,1098 @@
+
+[schema in depth guide]: https://github.com/diesel-rs/diesel/blob/master/guide_drafts/schema-in-depth.md
+
+
+## Database Models
+
+Now that we have a schema, Rust can interact with database tables,
+however that is a little different than interacting with a database model.
+Diesel is a bit different than other ORMS, in that its main focus is on query building.
+Coming from Ruby on Rails (or similar all-in-one frameworks),
+I expected to work with a single object as my database model,
+however in Diesel you should feel comfortable defining specialize structs to support your queries.
+This may seem odd at first, but it should feel more natural by the time we get to designing all CRUD endpoints.
+
+At the moment, we really only care about two database operations; Create and Read.
+This really just amounts to a SQL INSERT and SELECT statments.
+Diesel gives us some nice apis to load a whole row without SELECT,
+so we'll be using those.
+
+We have two database tables, users and posts, and we need to perform two operations on each.
+The data structs for our models *may* look like this.
+- User
+- NewUser
+- Post
+- NewPost
+
+Lets create a `src/models.rs` file and review the code that goes into it.
+
+
+```rust
+// Inside `src/models.rs`
+
+// We JUST made the schema file...
+// Lets take advantage of it by bringing it into scope here
+use schema::{posts, users};
+
+#[derive(Debug, Queryable)]
+struct User {
+    id: i32,
+    first_name: String,
+    last_name: String,
+    email: String,
+    password: String, 
+}
+
+#[derive(Debug, Insertable)]
+#[table_name="users"]
+struct NewUser {
+    first_name: String,
+    last_name: String,
+    email: String,
+    password: String, 
+}
+
+#[derive(Debug, Queryable)]
+struct Post {
+    id: i32,
+    user_id: i32,
+    title: String,
+    content: String,
+    published: bool,
+}
+
+#[derive(Debug, Insertable)]
+#[table_name="posts"]
+struct NewPost {
+    user_id: i32,
+    title: String,
+    content: String,
+}
+
+```
+
+If you're thinking *"whoa, whoa, whoa... WUT are those weird DERIVE things?!?!"*,
+don't worry, they're your friend!
+
+The `diesel_codegen` crate... well, generates code for us!
+Rust is strongly typed and has a wonderful trait system which Diesel takes advantage of.
+This gives us compile-time guarantees about all aspects of our models.
+
+`Debug` should look somewhat familiar, as it's basically standard with the Rust langauge.
+The really interesting ones are `Queryable` and `Insertable`.
+`Insertable` is Diesel's way of saying that *the values of this struct map to the columns of a table and can be inserted in a row*
+With `Insertable`, we don't generally include any auto-incrementing primary key columns because our database takes care of that for us.
+
+`Queryable` is also self explanatory.
+A `Queryable` struct is one that may be queried from the database with those fields / columns.
+Our `Queryable` structs share the same name as the table, so the table name is inferred,
```

Comment:
This is incorrect. Queryable does not care about the names of columns, or what table things come from. It cares about the order of the fields, and their types. But `Queryable` is by design decoupled from tables. You can have a `Queryable` struct that represents some random query that hits 5 different tables.